c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -3386,42 +3527,78 @@ class NativeCachingAllocator : public CUDAAllocator {
   // it will deleted in cudaIpcCloseMemHandle when its reference count is 0.
   //
   std::mutex IpcMutex;
-  ska::flat_hash_map<std::string, std::weak_ptr<void>> ipcMemHandle_to_devptr;
+  struct MemHandleCacheEntry {
+    MemHandleCacheEntry(
+        c10::DeviceIndex device,
+        std::string& handle,
+        const DeviceCachingAllocator& allocator)
+        : device_(device), cuda_ipc_ptr_(nullptr) {
+      std::istringstream ss(handle);
+      auto type = ss.get();
+      if (type == SHAREABLE_CUDA_MALLOC) {
+        cudaIpcMemHandle_t handle;
+        ss.read((char*)&handle, CUDA_IPC_HANDLE_SIZE);
+        C10_CUDA_CHECK(cudaIpcOpenMemHandle(
+            &cuda_ipc_ptr_, handle, cudaIpcMemLazyEnablePeerAccess));
+      } else if (type == SHAREABLE_CUDA_EXPANDABLE_SEGMENT) {
+        expandable_segment_ =
+            ExpandableSegment::fromShared(device, allocator.peers(), ss);
+      } else {
+        TORCH_INTERNAL_ASSERT(
+            false, "unexpected or illformed shareable handle type");
+      }
+    }
+    MemHandleCacheEntry(const MemHandleCacheEntry&) = delete;
+    MemHandleCacheEntry& operator=(const MemHandleCacheEntry&) = delete;
+    ~MemHandleCacheEntry() {
+      if (cuda_ipc_ptr_) {
+        cuda::CUDAGuard device_guard(device_);
+        C10_CUDA_CHECK(cudaIpcCloseMemHandle(cuda_ipc_ptr_));
+      }
```

Comment:
Just for my understanding, we don't have to close in the expandable segments case because it is already taken care of in fromShared: close((int)myfd); ?

Reply:
The expandable segment case is handled by the ExpandableSegment object dtor. We only need something special here because cuda_ipc_ptr_ is a raw object.